lib/graph_gateway.go

```diff
@@ -240,606 +217,1121 @@ func registerGraphGatewayMetrics() *GraphGatewayMetrics {
+func (it *backpressuredHAMTDirIterNoRecursion) Link() unixfs.LinkResult {
+	return it.curLnk
 }
 
-func (i *inboundBlockExchange) GetBlocks(ctx context.Context, cids []cid.Cid) (<-chan blocks.Block, error) {
-	mhMap := make(map[string]struct{})
-	for _, c := range cids {
-		mhMap[string(c.Hash())] = struct{}{}
+func (it *backpressuredHAMTDirIterNoRecursion) Next() bool {
+	defer func() {
+		if it.linksItr.Done() || it.err != nil {
+			if !it.hasClosed {
+				it.hasClosed = true
+				close(it.closed)
+			}
+		}
+	}()
+
+	if it.err != nil {
+		return false
 	}
-	mhs := make([]multihash.Multihash, 0, len(mhMap))
-	for k := range mhMap {
-		mhs = append(mhs, multihash.Multihash(k))
+
+	iter := it.linksItr
+	if iter.Done() {
+		return false
 	}
-	return i.ps.Subscribe(ctx, mhs...), nil
-}
 
-func (i *inboundBlockExchange) NotifyNewBlocks(ctx context.Context, blocks ...blocks.Block) error {
-	// TODO: handle context cancellation and/or blockage here
-	i.ps.Publish(blocks...)
-	return nil
+	/*
+		Since there is no way to make a graph request for part of a HAMT during errors we can either fill in the HAMT with
+		block requests, or we can re-request the HAMT and skip over the parts we already have.
+
+		Here we choose the latter, however in the event of a re-request we request the entity rather than the entire DAG as
+		a compromise between more requests and over-fetching data.
+	*/
+
+	var err error
+	for {
+		if it.ctx.Err() != nil {
+			it.err = it.ctx.Err()
+			return false
+		}
+
+		retry, processedErr := isRetryableError(err)
+		if !retry {
+			it.err = processedErr
+			return false
+		}
+
+		var nd ipld.Node
+		if err != nil {
+			var lsys *ipld.LinkSystem
+			lsys, err = it.getLsys(it.ctx, it.dirCid, gateway.CarParams{Scope: gateway.DagScopeEntity})
+			if err != nil {
+				continue
+			}
+
+			_, pbn, fieldData, _, _, ufsBaseErr := loadUnixFSBase(it.ctx, it.dirCid, nil, lsys)
+			if ufsBaseErr != nil {
+				err = ufsBaseErr
+				continue
+			}
+
+			nd, err = hamt.NewUnixFSHAMTShard(it.ctx, pbn, fieldData, lsys)
+			if err != nil {
+				err = fmt.Errorf("could not reify sharded directory: %w", err)
+				continue
+			}
+
+			iter = nd.MapIterator()
+			for i := 0; i < it.curProcessed; i++ {
+				_, _, err = iter.Next()
+				if err != nil {
+					continue
+				}
+			}
+
+			it.linksItr = iter
+		}
+
+		var k, v ipld.Node
+		k, v, err = iter.Next()
+		if err != nil {
+			retry, processedErr = isRetryableError(err)
+			if retry {
+				err = processedErr
+				continue
+			}
+			it.err = processedErr
+			return false
+		}
+
+		var name string
+		name, err = k.AsString()
+		if err != nil {
+			it.err = err
+			return false
+		}
+		var lnk ipld.Link
+		lnk, err = v.AsLink()
+		if err != nil {
+			it.err = err
+			return false
+		}
+
+		cl, ok := lnk.(cidlink.Link)
+		if !ok {
+			it.err = fmt.Errorf("link not a cidlink")
+			return false
+		}
+
+		c := cl.Cid
+
+		it.curLnk = unixfs.LinkResult{
+			Link: &format.Link{
+				Name: name,
+				Size: 0,
```

Comment:
Note: we're setting the size of elements in the dir_index.html listing to 0 for HAMT elements since we have no way of extracting TSize from the links while doing map iteration.

https://github.com/ipfs/go-unixfsnode/pull/51 was one attempt at resolving this which wasn't accepted due to being too hacky. IMO it'd probably be more natural to allow some typecasting here (e.g. typecast the iterator so that "Next" can return either an cidlink.Link or dag-pb Link).

cc @willscott

Reply:
calling `.Next()` gives you an `ipld.Node` currently, and you ask for it currently with `.AsLink()` to get the thing that you expect to be able to cast as `.(cidlink.Link)`.

okay if we have it be the node be the one we allow asking for the substrate of?

Reply:
my proposal for the upstream library is https://github.com/ipfs/go-unixfsnode/pull/60

Reply:
should be resolved now, thanks @willscott